aiida_vasp/utils/vasp/isif.py

```diff
@@ -0,0 +1,81 @@
+"""Helper for using ISIF INCAR tag."""
+import enum
+from aiida.common.utils import classproperty
+
+from aiida_vasp.utils.vasp.incar_param import AbstractIncarParam
+
+
+# pylint: disable=too-few-public-methods
+class IsifStressFlags(enum.Enum):
+    NONE = 0
+    FULL = 1
+    TRACE_ONLY = 2
+
+
+class Isif(AbstractIncarParam):
+    """
+    Helper for using ISIF INCAR tag, finds the correct value for a given set of degrees of freedom.
+
+    :param calculate_stress: IsifStressFlags.Full | TRACE_ONLY | NONE
+    :param vary_positions: bool, treat the ion positions as a degree of freedom?
+    :param vary_cell_shape: bool, treat the cell shape as a degree of freedom?
+    :param vary_cell_volume: bool, treat the cell volume as a degree of freedom?
+    """
+
+    ALLOWED_COMBINATIONS = {
+        (IsifStressFlags.NONE, True, False, False): 0,
+        (IsifStressFlags.TRACE_ONLY, True, False, False): 1,
+        (IsifStressFlags.FULL, True, False, False): 2,
+        (IsifStressFlags.FULL, True, True, True): 3,
+        (IsifStressFlags.FULL, True, True, False): 4,
+        (IsifStressFlags.FULL, False, True, False): 5,
+        (IsifStressFlags.FULL, False, False, False): 7,
+    }
+    POS_KEY = 'Positions'
+    SHAPE_KEY = 'Cell shape'
+    VOL_KEY = 'Cell volume'
+
+    def __init__(self, calculate_stress=IsifStressFlags.FULL, vary_positions=True, vary_cell_shape=False, vary_cell_volume=False):
+        if not isinstance(calculate_stress, IsifStressFlags):
+            raise ValueError('calculate_stress parameter must be of type {}'.format(IsifStressFlags))
+        self._stress = calculate_stress
+        self._dof = {self.POS_KEY: vary_positions, self.SHAPE_KEY: vary_cell_shape, self.VOL_KEY: vary_cell_volume}
+        self.validate()
+
+    @classproperty
+    def name(self):
+        return 'ISIF'
+
+    @property
+    def info(self):
+        msg = 'Calculate stress tensor: {stress} | DoF: {dof} | fixed: {fixed}'
+        stress = self._stress.name.capitalize()
+        dof = ', '.join([key for key, value in self._dof.items() if value])
+        fixed = ', '.join([key for key, value in self._dof.items() if not value])
+        return msg.format(stress=stress, dof=dof, fixed=fixed)
+
+    def _combination(self):
+        return (self._stress, self._dof[self.POS_KEY], self._dof[self.SHAPE_KEY], self._dof[self.VOL_KEY])
+
+    def validate(self):
+        if self._combination() not in self.ALLOWED_COMBINATIONS:
+            raise ValueError('Vasp does not allow the following combination of DoF: {}'.format(self.info))
+
+    def clean(self, incar_dict):
+        errors = []
+        warnings = []
+        if self._dof[self.VOL_KEY]:
+            if 'encut' not in incar_dict:
+                warnings.append(
+                    'ENCUT was not given (should be set >= 1.3xmax(ENMAX)) when running cell volume relaxation ({})'.format(self))
```

Comment:
The message is just a warning, so it does not stop you from running with something else and any workchain can choose to discard it or append it to the log.

I can agree to dropping the default behaviour for the relax workchain to set ENCUT = 1.3 * max(ENMAX), but I would like to leave the Encut class in for users that would want to set ENCUT based on ENMAX in order to be consistent with their current way of working. Even if it may not always be a good idea, I would leave this up to users.

Reply:
But the warning should be useful to the users, otherwise it is misleading. In fact this statement in the manual of 1.3 is just there mainly for historical reasons. I think the `ENCUT` class has to live on its own as we need this all over. In fact I think we need to move these tags into its own datastructure as I mentioned in another comment. Question is...we already have `parameters`  which is supposed to be named `incar` at some point. Do we then create some kind of `WorkingIncar` which is a class that resides from the start and takes all necessary modifications we do in the workchains? We need to preserve the `incar` and then we get an `output_incar` from these modification. All entries we modify goes into `outcar_incar` by default, regardless of what the parser does to OUTCAR or vasprun.xml at the end of the run. I am happy to leave this and put it on the issue list so that we get this workchain out asap.

Reply:
I will vote for leaving all the inputs and outputs unchanged. Instead, we should provide an incar builder which can optionally be used to build an incar dictionary safely. This can then be stored in a ParameterData and passed around as needed. That way we don't make life more complicated for expert users.

Reply:
Yes, I support this. But we need the output tags used so that we can compare and track it between the different workchains.